Add an experimental quirk for Facebook live streaming feature

youennf opened the pull request:
#### 91bd98d1d80307f26c10dbf959a93af7d32df0dc
<pre>
Add an experimental quirk for Facebook live streaming feature
<a href="https://rdar.apple.com/41104397">rdar://41104397</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298253">https://bugs.webkit.org/show_bug.cgi?id=298253</a>

Reviewed by Chris Dumez.

Live streaming is enabled by a feature flag.
We enable this flag using the web page mechanism to generate the list of feature flags.
The web page builds this list by grabbing specific script elements using a query.
We add to the list returned by the query a synthesized script element with just the flag enabled by default.
This is controlled by a newly added quirk.

Manually tested.

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::querySelectorAll):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFacebookFlagQuirk const):
(WebCore::createFacebookFlagElement):
(WebCore::copyElements):
(WebCore::Quirks::applyFacebookFlagQuirk):
(WebCore::handleFacebookQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/299802@main">https://commits.webkit.org/299802@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6aa1cba851b9eea9a07ddb94fc519b8893871301

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/120224 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/30569 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/126580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/72297 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/122100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/48495 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/91299 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60599 "Passed tests") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/123176 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/32436 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107786 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71853 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/31469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25892 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/70206 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101916 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/26077 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/129478 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/47144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/35768 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99920 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/47510 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103971 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99762 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25331 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/45215 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/23242 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43775 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/47007 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52712 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/46474 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49821 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/48158 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1146,6 +1148,29 @@ ExceptionOr<Element*> ContainerNode::querySelector(const String& selectors)
+static Vector<Ref<Element>> copyElements(NodeList& nodeList)
+{
+    Vector<Ref<Element>> elements;
+    for (size_t cptr = 0; cptr < nodeList.length(); ++cptr) {
```
Can we use `i` as usually instead of `cptr`?

cdumez commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1146,6 +1148,29 @@ ExceptionOr<Element*> ContainerNode::querySelector(const String& selectors)
+    return script;
+}
+
+static Vector<Ref<Element>> copyElements(NodeList& nodeList)
```
Can probably use `const NodeList&`

youennf replied:
ok

cdumez commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1146,6 +1148,29 @@ ExceptionOr<Element*> ContainerNode::querySelector(const String& selectors)
 }
 
+#if ENABLE(MEDIA_STREAM)
+static Ref<Element> createFlagElement(Document& document, ASCIILiteral value)
```
This seems very specific to Facebook and specific to the new Quirk. I think it would be better if this code lived in Quirks.cpp, if if the name mentioned Facebook.

youennf replied:
ok

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -906,6 +909,42 @@ bool Quirks::shouldMakeEventListenerPassive(const EventTarget& eventTarget, cons
+static Vector<Ref<Element>> copyElements(const NodeList& nodeList)
+{
+    Vector<Ref<Element>> elements;
+    for (size_t cptr = 0; cptr < nodeList.length(); ++cptr) {
```
Could we use `I` instead of `cptr`?

youennf committed:
Add an experimental quirk for Facebook live streaming feature
rdar://41104397
https://bugs.webkit.org/show_bug.cgi?id=298253

Reviewed by Chris Dumez.

Live streaming is enabled by a feature flag.
We enable this flag using the web page mechanism to generate the list of feature flags.
The web page builds this list by grabbing specific script elements using a query.
We add to the list returned by the query a synthesized script element with just the flag enabled by default.
This is controlled by a newly added quirk.

Manually tested.

* Source/WebCore/dom/ContainerNode.cpp:
(WebCore::ContainerNode::querySelectorAll):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldEnableFacebookFlagQuirk const):
(WebCore::createFacebookFlagElement):
(WebCore::copyElements):
(WebCore::Quirks::applyFacebookFlagQuirk):
(WebCore::handleFacebookQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/299802@main
